[libpas] Use page-based-zero-clear more aggressively

Constellation opened the pull request:
#### 6cba932c7d4f99bba227fb58ff1d7b42546d3cf8
<pre>
[libpas] Use page-based-zero-clear more aggressively
<a href="https://bugs.webkit.org/show_bug.cgi?id=301819">https://bugs.webkit.org/show_bug.cgi?id=301819</a>
<a href="https://rdar.apple.com/163879917">rdar://163879917</a>

Reviewed by Mark Lam.

Previously we were failing to use page-based-zero-clear when the first
address is not aligned to page size. This patch relaxes this condition.
We should use page-based-zero-clear for the middle of the allocated
region.

* Source/bmalloc/libpas/src/libpas/pas_allocation_result.c:
(pas_allocation_result_zero_large_slow):

Canonical link: <a href="https://commits.webkit.org/302466@main">https://commits.webkit.org/302466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/511974f1fe4d3e0a01efcd6d464a05496ee07590

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39964 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136507 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80511 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ca48bf1a-52e6-4123-94ac-eb03662b48dc) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1264 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98320 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66197 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/524b57f3-afa6-41cb-9a1f-58c975f1f54e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132075 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1025 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78965 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ef96e473-75ea-48a2-9461-616c98f095fd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/948 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33787 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79786 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121120 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109395 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34283 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138981 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127580 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1180 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1143 "Found 1 new test failure: fast/images/individual-animation-toggle.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106855 "Found 1 new test failure: http/wpt/misc/no-last-modified.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112006 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106684 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/967 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30530 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53744 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20169 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1253 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160595 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1079 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40085 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1124 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1177 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/bmalloc/libpas/src/libpas/pas_allocation_result.c:
```diff
@@ -35,15 +35,30 @@
+            pas_zero_memory((void*)page_aligned_end, end - page_aligned_end);
+    } else
+        pas_zero_memory((void*)begin, size);
+
```
I can't understand much of the libpas code, but I can definitely follow this PR!

A side-effect of this PR is that we'll be using [madvise(..., MADV_FREE)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/madvise.2.html) (called by `pas_page_malloc_zero_fill`) more where supported :)

I see that `pas_page_malloc_alignment` is also called in `pas_page_malloc_zero_fill` (so we could pass it in to avoid this), but I'm guessing the cost here is negligible.

LGTM

MenloDorian reviewed (approved):
r=me

Constellation committed:
[libpas] Use page-based-zero-clear more aggressively
https://bugs.webkit.org/show_bug.cgi?id=301819
rdar://163879917

Reviewed by Mark Lam.

Previously we were failing to use page-based-zero-clear when the first
address is not aligned to page size. This patch relaxes this condition.
We should use page-based-zero-clear for the middle of the allocated
region.

* Source/bmalloc/libpas/src/libpas/pas_allocation_result.c:
(pas_allocation_result_zero_large_slow):

Canonical link: https://commits.webkit.org/302466@main
